Core/include/Acts/Propagator/MultiEigenStepperLoop.hpp

```diff
@@ -0,0 +1,800 @@
+// This file is part of the Acts project.
+//
+// Copyright (C) 2021 CERN for the benefit of the Acts project
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+#pragma once
+
+// Workaround for building on clang+libstdc++
+#include "Acts/Utilities/detail/ReferenceWrapperAnyCompat.hpp"
+
+#include "Acts/Definitions/Algebra.hpp"
+#include "Acts/Definitions/Units.hpp"
+#include "Acts/EventData/MultiComponentBoundTrackParameters.hpp"
+#include "Acts/EventData/TrackParameters.hpp"
+#include "Acts/MagneticField/MagneticFieldProvider.hpp"
+#include "Acts/Propagator/EigenStepper.hpp"
+#include "Acts/Propagator/EigenStepperError.hpp"
+#include "Acts/Propagator/Propagator.hpp"
+#include "Acts/Utilities/Intersection.hpp"
+#include "Acts/Utilities/Result.hpp"
+#include "Acts/Utilities/detail/gaussian_mixture_helpers.hpp"
+
+#include <algorithm>
+#include <cmath>
+#include <functional>
+#include <limits>
+#include <numeric>
+#include <sstream>
+#include <vector>
+
+#include <boost/container/small_vector.hpp>
+
+#include "MultiStepperError.hpp"
+
+namespace Acts {
+
+using namespace Acts::UnitLiterals;
+
+/// @brief Reducer struct for the Loop MultiEigenStepper which reduces the
+/// multicomponent state to simply by summing the weighted values
+struct WeightedComponentReducerLoop {
+  template <typename component_range_t>
+  static Vector3 toVector3(const component_range_t& comps,
+                           const FreeIndices i) {
+    return std::accumulate(
+        comps.begin(), comps.end(), Vector3{Vector3::Zero()},
+        [i](const auto& sum, const auto& cmp) -> Vector3 {
+          return sum + cmp.weight * cmp.state.pars.template segment<3>(i);
+        });
+  }
+
+  template <typename stepper_state_t>
+  static Vector3 position(const stepper_state_t& s) {
+    return toVector3(s.components, eFreePos0);
+  }
+
+  template <typename stepper_state_t>
+  static Vector3 direction(const stepper_state_t& s) {
+    return toVector3(s.components, eFreeDir0).normalized();
+  }
+
+  template <typename stepper_state_t>
+  static ActsScalar momentum(const stepper_state_t& s) {
+    return std::accumulate(
+        s.components.begin(), s.components.end(), ActsScalar{0.},
+        [](const auto& sum, const auto& cmp) -> ActsScalar {
+          return sum +
+                 cmp.weight * (1 / (cmp.state.pars[eFreeQOverP] / cmp.state.q));
+        });
+  }
```

Comment:
Se my comment from before - do you have a feeling if we could gain from something like this?

Reply:
As mentioned above: Yes this could potentially reduce both code size and increase performance here...

However, I'm not sure if I would implement this immediately or do a big performance measurement some time and keep it in mind...

Reply:
So to resume on this:

You mean something like this, right? (taken arbitrarily from Navigator.hpp:366)

```c++
state.navigation.currentVolume = boundary->attachedVolume(
     state.geoContext, stepper.position(state.stepping),
     stepper.direction(state.stepping), state.stepping.navDir);
```

And here do something like:

```c++

const auto [pos, dir] = stepper.posdir(state.stepping);

state.navigation.currentVolume = boundary->attachedVolume(
     state.geoContext, pos, dir, state.stepping.navDir);
```

But this would change the whole stepper-navigator-interaction, right? Or do I misunderstand what your idea is?

Reply:
We can leave this as an optimisation step for later.